### PR TITLE
fix: expand assoc types to records and erase assoc types throughout expressions in Monomorph

### PR DESCRIPTION
**Pull request by mlutze**
related to #5647. There is a performance risk here: we perform AssocType resolution on all types in Monomorph

### CODE REVIEW

**Comment by magnus-madsen** [comment]
What's the output of `--Xbenchmark-phases` then?

**Comment by mlutze** [comment]
```
====================== Flix Compiler Phases ======================              

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Deriver                          3.6 ms (00.0%)
  Tailrec                         12.9 ms (00.1%)
  Reader                          16.8 ms (00.1%)
  Regions                         31.6 ms (00.3%)
  VarNumbering                    33.5 ms (00.3%)
  Safety                          48.8 ms (00.4%)
  Simplifier                      53.1 ms (00.4%)
  PatternExhaustiveness           68.9 ms (00.6%)
  LambdaLift                      73.1 ms (00.6%)
  Eraser                          84.3 ms (00.7%)
  EarlyTreeShaker                 84.6 ms (00.7%)
  ClosureConv                    118.9 ms (01.0%)
  LateTreeShaker                 121.2 ms (01.0%)
  Stratifier                     145.8 ms (01.2%)
  Namer                          199.7 ms (01.6%)
  Kinder                         201.1 ms (01.6%)
  Lowering                       214.0 ms (01.7%)
  Finalize                       235.5 ms (01.9%)
  Weeder                         264.9 ms (02.1%)
  Instances                      267.8 ms (02.1%)
  Redundancy                     314.6 ms (02.5%)
  Resolver                       589.9 ms (04.7%)
  Optimizer                      757.2 ms (06.1%)
  Parser                         764.9 ms (06.1%)
  Monomorph                      1394.4 ms (11.1%)
  JvmBackend                     2131.4 ms (17.0%)
  Typer                          4280.8 ms (34.2%)
```

**Comment by magnus-madsen** [comment]
What's the delta?

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala [hunk 1]
```diff
@@ -255,65 +255,72 @@ object Monomorph {
     */
   private def specialize(exp0: Expression, env0: Map[Symbol.VarSym, Symbol.VarSym], subst0: StrictSubstitution, def2def: Def2Def, defQueue: DefQueue)(implicit root: Root, flix: Flix): Expression = {
 
+    /**
+      * Specializes the given type.
+      */
+    def visitType(t: Type, subst: StrictSubstitution): Type = {
+      EqualityEnvironment.reduceType(subst(t), root.eqEnv).get
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala:262]: This probably visit the type twice, right?

**Comment by magnus-madsen** [comment]
The cost seems pretty steep. I see a 2x increase.

**Comment by mlutze** [comment]
:(

**Comment by mlutze** [comment]
I think it can instead be done during type reconstruction.

**Comment by mlutze** [comment]
Cost is now around 10-15% increase in Monomorph time.

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
> Cost is now around 10-15% increase in Monomorph time.

That's fine. I can try locally too.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**